includes/widgets/class-amp-widget-categories.php

```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * Class AMP_Widget_Categories
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Widget_Categories
+ *
+ * @package AMP
+ */
+class AMP_Widget_Categories extends WP_Widget_Categories {
+
+	/**
+	 * Echoes the markup of the widget.
+	 *
+	 * @todo filter $output, to strip the <script> tag.
+	 * @see https://github.com/Automattic/amp-wp/issues/864
+	 * @param array $args Widget display data.
+	 * @param array $instance Data for widget.
+	 * @return void.
+	 */
+	public function widget( $args, $instance ) {
+		ob_start();
+		parent::widget( $args, $instance );
+		$output = ob_get_clean();
```

Comment:
@kienstra Here we need to handle the dropdown behavior. I think this should strip out the `script` tag here via regex:

```php
$output = preg_replace( '#<script.+?</script>#', '', $output );
```

And then what it needs to do is inject two additional things:

1. Add am `id` attribute to the `<form>` with a random value, or an auto-incremented one: `widget-categories-dropdown-%d`, e.g. `widget-categories-dropdown-123`
2. Inject an `on` attribute onto the `<select>` element like `on="change:widget-categories-dropdown-123.submit"`.

Reference: https://www.ampproject.org/docs/reference/components/amp-form#input-events

👉 N.B. WordPress 4.9 first introduced the `<form>` element here: https://github.com/WordPress/wordpress-develop/commit/b7c70cad148a5a4c57b8403ed4ba5008f821a67a

So if no `<form>` is present, then it should wrap the `<select>` with one.

On second thought, instead of output-buffering the output and modifying it, just copy the existing logic from `\WP_Widget_Categories::widget()` in WP 4.9 and modify it here to be valid AMP.

Reply:
**Applied Suggestion**
*Thanks For The Explanation*

Hi @westonruter,
Thanks for your great explanation of how to make the dropdown work with AMP. How does [this commit](https://github.com/Automattic/amp-wp/pull/870/commits/fb9b451c2286f52b3a35766c52525fc5ff167868) look? There are still Travis errors that I need to fix.

It mainly copies  `\WP_Widget_Categories::widget()` from 4.9, and makes the changes you suggested. 

It's worked locally, with valid AMP.